### DISCUSSION TITLE
How does the operation cache interact with foreign libraries?

### DISCUSSION THREAD

**Discussion by mertalev** [libvips/libvips#3823] [Q&A] [answered]
I noticed that when processing RAW images that get forwarded to ImageMagick, memory doesn't seem to be released afterwards. This can balloon to many gigabytes with additional images. Setting `sharp.cache({ files: 0 });` makes the RAM usage much more consistent.

Since there's also a `memory` limit for the cache that defaults to 50mb, I'm not sure how that fits into this. My guess is that `memory` only tracks the memory used internally in libvips but not in ImageMagick, and holding file descriptors open somehow prevents releasing memory. Does that sound right?

For reference, this is with libvips 8.15 and ImageMagick 7.1.1-24.

**Comment by jcupitt** [comment, marked as answer]
Hi @mertalev,

At the end of evaluation, all objects in a pipeline are sent a `minimise` signal. What they do varies, but the idea is to drop caches, free file descriptors, and so on. Anything that can be recalculated if evaluation happens again is supposed to go. Imagemagick doesn't support this sort of stop/restart, so the magick loader ignores the signal, unfortunately. 

As you say, this means that a lot of resources can get stuck inside magick load objects. The libvips operation cache drops operations in an LRU manner to get cache usage inside the set limits, but it can only track memory and file descriptors allocated by libvips. So again, large magick loaders can be missed. I usually try to avoid using imagemagick for large images, though I realize this is not always possible.

In your case, the best solution would probably be to add a native `rawload` operation to libvips which could support `minimise`. This has been talked about before, but no one's got around to doing the work :( Part of the problem is there has not been a good libraw for libvips to wrap.

You probably know, but im6 usually uses about half the memory of im7. Downgrading might help a bit with your memory issues.

**Reply by mertalev** [reply]
Thanks so much for the very detailed explanation! That all makes sense.

Having a `rawload` integrated with libraw would be amazing — we could potentially drop to just libvips and libraw. But I also totally relate to wanting to add a feature, but not having the time or bandwidth for it.

**Comment by jcupitt** [comment]
See also https://github.com/libvips/libvips/discussions/2772
